### DISCUSSION TITLE
How to partition heterogeneous elastic properties vectors in different processes?

### DISCUSSION THREAD

**Discussion by rafaelmarch3** [mfem/mfem#2469] [Q&A] [answered]
Hello all, 

My objective is to determine effective elastic properties of heterogenous 3D materials.

I have successfully adapted Example 2 to handle the elastic deformation of an arbitrary 3D cartesian mesh with heterogeneous elastic property fields (i.e. one mu and lambda for each cell). The serial version of this code works fine. It uses PWConstCoefficient to define the elastic properties. 

I'm now working on the parallel version of this code, but I cannot figure out how to populate the lambda and mu coefficient vectors for each process:

For instance if my volume is 30 x 30 x 30 and I have 3 processes:

1) I read an input file that contains 27,000 lambda and mu values. Note that my problem is generally not too large so I let all processes read these input data arrays and store it completely in their memories;

2) Now if I check pmesh->attributes.Max(), for each process (pmesh being an instance of ParMesh), it has less elements than 27,000, since each process stores only part of the domain;

3) How to map the input lambda and mu values to the local localLambda and localMu arrays, respecting MFEM's partitioning? Is there any utility function to achieve this?

```
// Setting mesh lambda and mu
Vector localLambda(pmesh->attributes.Max());
Vector localMu(pmesh->attributes.Max());
for(auto i=0; i<pmesh->attributes.Max(); i++)
{
    localLambda(i) = ??;
    localMu(i) = ??;
}
```

Thanks, 
Rafael.

**Comment by pazner** [comment, marked as answer]
Hi @rafaelmarch3,

If I understand correctly, you read from your input file an array `globalLambda` of size 27,000, where `globalLambda[i-1]` is the lambda value corresponding to the `i`th attribute, and similarly for `globalMu`.

If this is the case, then I believe that on each MPI rank, you can construct the `localLambda` and `localMu` arrays as follows
```c++
Vector localLambda(pmesh->attributes.Max());
Vector localMu(pmesh->attributes.Max());
localLambda = 0.0;
localMu = 0.0; // Initialize to zero: not strictly necessary, but makes sure these arrays have no un-initialized values
for (int attribute : pmesh->attributes)
{
   localLambda[attribute - 1] = globalLambda[attribute - 1];
   localMu[attribute - 1] = globalMu[attribute - 1];
}
```

**Reply by rafaelmarch3** [reply]
Hello, many thanks for your reply. Indeed, I guess the key was to set the attributes before creating the parallel mesh. This ensures that all the material attributes are partitioned accordingly. Then we can set the lambda e mu vectors depending on how many attribute labels (materials) we have. Thanks again!
